pkgs/stdenv/linux/default.nix

```diff
@@ -9,14 +9,11 @@
 # is used to build all other packages (including the bootstrapFiles).
 #
 # Goals of the bootstrap process:
```

Comment:
@trofi, did you try splitting `gcc`'s stage1 and stage2 into separate derivations?  That would achieve the goal without having to build `gcc` twice.  But it is a larger project than simply adding an extra bootstrap stage, so if there is a deadline or urgent need to deal with this...

Reply:
I plan to explore `--disable-bootstrap` for `gcc` builds when `stdenv.cc.cc` and built `gcc` are of the same versions. But there are extra caveats to consider that might break non-bootstrap form of it. That should shave off 0.5 of gcc build time.

Splitting `gcc`'s build system into a multiple derivations is a very big project. Looking at a smaller scale `binutils` / `libbfd` split `nixpkgs` has I would say it's a maintenance nightmare. Testing it in bootstrap is another level of footgun. Adding support for it upstream would make me less nervous :)

Reply:
I already suspected that messing with gcc's stages could help, perhaps skipping some of the stages when we "build twice" anyway, but I don't know gcc's build process really so won't be able to help much.  I certainly wouldn't attempt such big projects inside this PR; better leave them for later.

Reply:
I don't understand what the big rush is.

This PR makes it more difficult to fix the problem the right way later on.

Reply:
The "big hurry" is that we are encountering more and more packages that don't build on aarch64, which is stuck on GCC 9 by default due to the libgcc_s issue.

Reply:
> aarch64, which is stuck on GCC 9 by default due to the libgcc_s issue.

Is there an issue or link for this?  The holdback is in `all-packages.nix`, which takes forever to run `git blame` on (and many commits have touched that part).

I'm trying to reproduce the problem but I only have fast builders for `powerpc64le` and `x86_64`.

I'm pretty sure there is a much less invasive workaround.

If this truly is the reason for urgency (and PRs that copy-and-paste code their author admits to not understanding!) that should be mentioned prominently in the commit message.

Reply:
Alright after seven minute-long `git blame`s I found the issue:

https://github.com/NixOS/nixpkgs/issues/108111

The commit message for this PR needs to give this background.

Reply:
I'm having a hard time believing that this is the reason for the urgency.

If this was the reason for such a rush, we could simply [update the bootstrap-files](https://github.com/NixOS/nixpkgs/pull/207135#issuecomment-1362225258) instead.

That's not a long-term solution, but the copy-and-pasting business in this PR, and imposing the rebuild costs on all platforms (instead of aarch64) isn't a long-term solution either.

Rebuilding the aarch64 bootstrap-files is an easy-to-revert change.  This PR is not.

Reply:
> This PR makes it more difficult to fix the problem the right way later on.

Why do you think it will make things more difficult to fix?

You can just remove one extra stage (and maybe shuffle around a few thing you don't want to rebuild) and add your steps. In fact if you were to change `gcc` outputs you could just throw all the stages out and add new ones. It's trivial if you know you want to achieve.

I'll cite the things to roll back here in full:

```nix 
  # 2nd stdenv that contains our own rebuilt binutils and is used for
  # compiling our own gcc.
  #
  # resulting stage2 stdenv:
  # - coreutils, glibc: from bootstrapFiles
  # - binutils, gcc: from nixpkgs, built by bootstrapFiles toolchain
  (prevStage: stageFun prevStage {
    name = "bootstrap-stage2";

    overrides = self: super: {
      inherit (prevStage)
        ccWrapperStdenv
        bash bison gnum4 coreutils gnugrep
        patchelf perl texinfo xz;

      ${localSystem.libc} = getLibc prevStage;

      gcc-unwrapped =
        let makeStaticLibrariesAndMark = pkg:
              lib.makeOverridable (pkg.override { stdenv = self.makeStaticLibraries self.stdenv; })
                .overrideAttrs (a: { pname = "${a.pname}-stage2"; });
        in super.gcc-unwrapped.override {
        # Link GCC statically against GMP etc.  This makes sense because
        # these builds of the libraries are only used by GCC, so it
        # reduces the size of the stdenv closure.
        gmp = makeStaticLibrariesAndMark super.gmp;
        mpfr = makeStaticLibrariesAndMark super.mpfr;
        libmpc = makeStaticLibrariesAndMark super.libmpc;
        isl = makeStaticLibrariesAndMark super.isl_0_20;
        # Use a deterministically built compiler
        # see https://github.com/NixOS/nixpkgs/issues/108475 for context
        reproducibleBuild = true;
        profiledCompiler = false;
      };
    };

    # `libtool` comes with obsolete config.sub/config.guess that don't recognize Risc-V.
    extraNativeBuildInputs =
      lib.optional (localSystem.isRiscV) prevStage.updateAutotoolsGnuConfigScriptsHook;
  })
```

39 lines of well isolated change.

The other changes do not need to be rolled back and are quality-of-life improvements. They are prerequisites for this change to work though.

> If this was the reason for such a rush, we could simply https://github.com/NixOS/nixpkgs/pull/207135#issuecomment-1362225258 instead.

Every time new person encountered `libgcc_s.so.1` mismatch it was proposed to update `bootstrapTools`. It never happened. I remember it ongoing for at least 6 months. Looking at `gcc` versions I think we are broken for a year.

> Is there an issue or link for this?

There are many issues:

- One of the more comprehensive ones is linked under this PR: https://github.com/NixOS/nixpkgs/issues/208412
- Another is that `nixpkgs` `glibc` is built using `bootstrapTools`'s `gcc`. If we don't upgrade `gcc` on a regular basis for all targets (and we don't, and I don't see us doing it) it's a very annoying property to keep in mind.

> Rebuilding the aarch64 bootstrap-files is an easy-to-revert change. This PR is not.

Mechanically they are equally easy to revert. Breakage-wise they are both hard to revert as dependencies rely on environment correctness.

Reply:
> The other changes do not need to be rolled back and are quality-of-life improvements.

Great!  Please break those out into a standalone PR and let's get them merged.  I have no problems with any of those.  I would be happy to immediately approve a PR with those three commits.

> 39 lines of well isolated change.

No, they're not well-isolated.  Also, 6e67eab46ff503bbb99bf51801dd0cd9c4027ae8 is more than 39 lines of changes.

It's very easy for external code to become dependent upon the extra stage due to laziness and evaluation order.  Since you're making this change *globally* across all non-aarch64 platforms nobody will be using the pre-PR codepath, so we will never know if it breaks.  This is my main objection.

If 6e67eab46ff503bbb99bf51801dd0cd9c4027ae8 is truly "well isolated" it should be easy to make them conditional and enabled **only on aarch64**.  This way we will be constantly testing our ability to revert this change (every x86_64 rebuild will verify that).  If you did that -- make all the changes in 6e67eab46ff503bbb99bf51801dd0cd9c4027ae8 conditional on `isAarch64` -- I would would be a lot less worried about this approach.

My smaller concern is the fact that you're copy-and-pasting large blocks of code (which you [admit you don't understand!](https://github.com/NixOS/nixpkgs/pull/209063/commits/6e67eab46ff503bbb99bf51801dd0cd9c4027ae8#r1063095582)) instead of at the very least factoring out the common expression.  I mean... come on, do you want to work on a codebase where people do stuff like that?  Nixpkgs is a pleasure to work on because it doesn't have stuff like that.

Reply:
> Great! Please break those out into a standalone PR and let's get them merged.

They were broken out. The list is at https://github.com/NixOS/nixpkgs/issues/208412#issuecomment-1371211147

> No, they're not well-isolated.

Can you be more specific? What are the dependencies that will require you to change things you won't be able to do perform or easily diagnose?

> Also, https://github.com/NixOS/nixpkgs/commit/6e67eab46ff503bbb99bf51801dd0cd9c4027ae8 is more than 39 lines of changes.

39 lines needs a roll back. I can make it 39 lines and reintroduce back trivial rebuilds for `xz` and friends. They are trivial fixes though, I don't think it's necessary, but I can make redundancy removal a separate PR as well.

> It's very easy for external code to become dependent upon the existence of the extra stage due to laziness and evaluation order.

Can you give more specific of dependency you have in mind and why it will be hard to fix? In any case I will give my help and will spend the time to fix those issues in the future.

> If https://github.com/NixOS/nixpkgs/commit/6e67eab46ff503bbb99bf51801dd0cd9c4027ae8 is truly "well isolated" it should be easy to make them conditional and enabled only on aarch64.

I want every arch to benefit from `libgcc_s.so.1` to match `gcc` version and have `glibc` built by non-bootstrap compiler.  I very specifically sent out https://github.com/NixOS/nixpkgs/pull/208478 to get a consensus that state of today's `glibc` is a violation of the principles. This PR fixes it. aarch64 linakge fix is a natural outcome of it.

> My smaller concern is the fact that you're copy-and-pasting large blocks of code (which you [admit you don't understand!](https://github.com/NixOS/nixpkgs/pull/209063/commits/6e67eab46ff503bbb99bf51801dd0cd9c4027ae8#r1063095582)) instead of at the very least factoring out the common expression. 

I don't understand the implications of removing static library overrides and would like to clean them up separately. I have vague understanding why they are there. I'll reiterate again: I will spend time to clean them up when/if this PR is merged. Switching from shared to static is not a small task as dynamic linking during stdenv bootstrap is very fragile due to incorrect library lookup paths. I expect to spend a lot of time on it. Moving `libstdc++.so` (the part if this  PR) is a clear example of that. It will need to be done for every gcc library it uses. I think this PR is big enough on it's own.

> I mean... come on, do you want to work on a codebase where people do stuff like that?

I am not just willing to work on that. I want to improve it in very short term future. I don't think that trivial copy if a `gcc` attribute is the problem for those who work on the bootstrap of linux for hacking. I want to think this PR shows it very clearly.

> Nixpkgs is a pleasure to work on because it doesn't have any of that. Maybe in the "leaf packages", but definitely not in stdenv or lib.

I would like to think this PR would not inhibit anyone to work on bootstrap stdenv and would not make it less straightforward to reason about. If I read between the lines this PR makes something harder for you to work on. Can you, maybe, elaborate? Or is your concern only extra overhead it pulls in?